src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -0,0 +1,559 @@
+    {
+        ChipLogError(Zcl, "DEM: Failed to PowerAdjustRequest() args.");
+    }
+}
+
+void Instance::HandleCancelPowerAdjustRequest(HandlerContext & ctx,
+                                              const Commands::CancelPowerAdjustRequest::DecodableType & commandData)
+{
+    Status status = Status::Success;
+    ESAStateEnum esaStatus;
+
+    /* Check that the ESA state is PowerAdjustActive */
+    esaStatus = mDelegate.GetESAState();
+    if (ESAStateEnum::kPowerAdjustActive != esaStatus)
+    {
+        ChipLogError(Zcl, "DEM: kPowerAdjustActive != esaStatus");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    status = mDelegate.CancelPowerAdjustRequest();
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    if (status != Status::Success)
+    {
+        ChipLogError(Zcl, "DEM: Failed to CancelPowerAdjustRequest()");
+        return;
+    }
+}
+
+void Instance::HandleStartTimeAdjustRequest(HandlerContext & ctx,
+                                            const Commands::StartTimeAdjustRequest::DecodableType & commandData)
+{
+    Status status                    = Status::Success;
+    uint32_t earliestStartTimeEpoch  = 0;
+    uint32_t latestEndTimeEpoch      = 0;
+    uint32_t requestedStartTimeEpoch = commandData.requestedStartTime;
+    uint32_t duration;
+
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecastNullable = mDelegate.GetForecast();
+
+    if (forecastNullable.IsNull())
+    {
+        ChipLogError(Zcl, "DEM: Forecast is Null");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    if (ESAStateEnum::kUserOptOut == mDelegate.GetESAState())
+    {
+        ChipLogError(Zcl, "DEM: ESAState = kUserOptOut");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    /* Temporary variable to save calling .Value() on forecastNullable */
+    auto & forecast = forecastNullable.Value();
+
+    /**
+     * If the RequestedStartTime value resulted in a time shift which is
+     * outside the time constraints of EarliestStartTime and
+     * LatestEndTime, then the command SHALL be rejected with CONSTRAINT_ERROR;
+     * in other failure scenarios the command SHALL be rejected with FAILURE
+     */
+    /* earliestStartTime is optional based on the StartTimeAdjust (STA) feature AND is nullable */
+    if (!(forecast.earliestStartTime.HasValue()) || !(forecast.latestEndTime.HasValue()))
+    {
+        /* These should not be NULL since this command requires STA feature and these are mandatory for that */
+        ChipLogError(Zcl, "DEM: EarliestStartTime / LatestEndTime do not have values");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    /* Temporary variable to save keep calling .Value() on the Optional element */
+    auto & earliestStartTimeOptional = forecast.earliestStartTime.Value();
```

Comment:
```suggestion
    auto & earliestStartTimeNullable = forecast.earliestStartTime.Value();
```
would more accurately reflect what this is, right?

Reply:
The reason for calling it Optional was because  at this level it was the 'optional' wrapper of the inner 'nullable'.

Earliest start time is defined as:
Optional<DataModel::Nullable<uint32_t>> earliestStartTime;

So the outer ".Value()" call gets you to the Optional layer, and then calling .Value() (again) on the optional layer gets you to the Nullable layer.

At least that's my understanding of the class containers here...

Reply:
My point is `earliestStartTimeOptional` is not an optional value in any way.  Its type is `Nullable<uint32_t>`, no?  So calling it `...Optional` is really misleading.  It makes the reader think this variable is an `Optional<something>`, when it's not.

If you want to call it `earliestStartTimeNonOptional` that would more accurately reflect things.  Or `earliestStartTimeWithOptionalRemoved` or `earliestStartTimeWithOptionalUnwrapped`, but both of those are too long in my opinion.

If you had done `auto & earliestStartTimeOptional = forecast.earliestStartTime;`, _that_ would make sense.  But that's not what the code is doing here...

Reply:
I guess the original reason for saying it was optional was because it is only required if the FA feature is supported.
Both earliestStartTime and latestEndTime are optional based on FA feature.

However only the earliestStartTime can be NULL (implying it can start immediately).

Maybe I'll just rename the variable to something like temp ! :) and if I removed the auto then it might be clearer to the reader that it's got this Optional and Nullable...

Let me try something to see if it looks better.

Reply:
Made the suggested change